### PR TITLE
Remove chunks that do not intersect target area in gradient search resampling

### PR DESCRIPTION
**Pull request by pnuu**
This PR adds filtering to gradient search resampling that discards source data chunks that do not intersect the target area.

The gradient search is now an optional mode, and requires `shapely` to be installed.

With this PR the following work with gradient search resampling:

- resampling of datasets with different resolutions
- resampling of datasets that previously failed with `IndexError: Bad index` because chunks had incompatible shapes
- resampling of data with `AreaDefinition`
- resampling of data with `SwathDefinition` with "smooth" coordinates (= not VIIRS, MODIS, point data)

- [x] Tests added
- [x] Tests passed
- [x] Passes ``git diff origin/master **/*py | flake8 --diff``

### CODE REVIEW

**Comment by pnuu** [comment]
Some timing results so far for the following script:

```python
import glob
import os
os.environ['PYTROLL_CHUNK_SIZE'] = '1024'
os.environ['DASK_NUM_WORKERS'] = '1'
os.environ['OMP_NUM_THREADS'] = '1'

from satpy import Scene

fnames = glob.glob('/home/lahtinep/data/satellite/geo/msg/*202004170600*')
glbl = Scene(reader='seviri_l1b_hrit', filenames=fnames)
glbl.load(['natural_color', 'overview', 'red_snow'], generate=False)
lcl = glbl.resample('eurol', resampler='gradient_search', reduce_data=False)
lcl.save_datasets(base_dir='/tmp')
```

- `master` branch: 31.1 s (1 core) / 23.1 s (2 cores) / 20.6 s (4 cores)
- this pr: 23.6 s / 16.7 s / 14.7

This image shows how the data was reduced before resampling (black = target area, blue = chunks used in processing, red = discarded chunks):
geos_crop_boundaries_fixed_covers_seviri

There are still some problems with ABI data. Although the chunk filtering works ...
geos_crop_boundaries_fixed_covers

... the image still doesn't have full coverage:
airmass_20170727_090038

**Comment by pnuu** [comment]
The problem with ABI was my assumption of how the chunks are reordered for processing. The stepping goes first in Y direction, then in X. I've got it reversed. It works wit Seviri because there are only one chunk in X direction.

**Comment by djhoese** [comment]
> The problem with ABI was my assumption of how the chunks are reordered for processing. The stepping goes first in Y direction, then in X. I've got it reversed. It works wit Seviri because there are only one chunk in X direction.

So now after your last fix the ABI image is correct?

**Comment by pnuu** [comment]
The last commit fixed the chunk collection for ABI, so here are some timing results for ABI L1b (reduced resolution via Eumetcast) for three RGBs (`['airmass', 'ash', 'night_fog']`), resampled to my test area `'conus_5km'` (area def shown below):

- `master` branch: 30.8 s (1 core) / 24.5 s (2 cores) / 22.5 s (4 cores)
- this pr: 11.7 s / 9.8 s / 9.0 s

The six chunks that were used in processing are shown in the above image in blue surrounding the black target area.

```yaml
conus_5km:
  description: Continental USA, 5 km resolution
  projection:
    proj: laea
    a: 6370997.0
    b: 6370997.0
    lat_0: 50.0
    lon_0: -100.0
  shape:
    height: 669
    width: 1027
  area_extent:
    lower_left_xy: [-2864809.2037627744, -2830762.316377244]
    upper_right_xy: [2268528.5715566175, 514757.94726551295]
```

**Comment by pnuu** [comment]
> So now after your last fix the ABI image is correct?

Yes:
airmass_20170727_090038

**Comment by pnuu** [comment]
According to `dask.diagnostics.ResourceProfiler` the memory usage dropped from ~3.3 GB (master branch) to ~1.4 GB when using 4 Dask workers for the ABI case.

File: pyresample/gradient/__init__.py [hunk 1]
```diff
@@ -178,26 +310,45 @@ def reshape_to_stacked_3d(array):
...
-                       method=kwargs.get('method', 'bilinear'))
-    return da.nanmax(res, axis=-1).squeeze()
+    if data.ndim != 4:
+        raise NotImplementedError(
+            'Gradient search resampling only supports 4D arrays.')
+
+    if data.shape[-1] == 0:
+        logger.warning("Data doesn't cover the target area.")
+        return None
+    else:
```
Comment (djhoese) [inline pyresample/gradient/__init__.py:344]: The smallest of nit-picks, this `else` doesn't have to be here since the above `if` returns `None`.

**Comment by pnuu** [comment]
With this latest commit, the timings are

- ABI: 11. 3 s (1 core), 9.4 s (2 cores), 11.3 s (4 cores)
- SEVIRI: 24.3 s (1 core). 16.6 (2 cores), 14.7 s (4 cores)

For ABI there's a slight improvement even with this rather small target area, and with larger target area (= more chunks) the gain would be larger. There's zero gain for SEVIRI as all the used source chunks cover all the target chunks.

**Comment by pnuu** [comment]
Another test with 1 km CONUS area (3345 x 5135) and ABI:

- `master`: 5 m 11.5 s
- filter only source chunks: 1 m 7.3 s
- filter also target chunks (the latest commit): 1 m 9.4 s
- filter also target chunks, cache polygons: 1 m 7.5 s

Pretty significant jump from the `master` branch, I'd say.

To get the most of the target chunk filtering, we'd need to pre-compute all the boundary polygons, but I didn't find a way to pass them via `da.blockwise()` to the worker function. Or if there were even a way to pass some info on the chunk location it'd be possible to access them from each of the worker.

**Comment by pnuu** [comment]
After the last major restructuring (pair-up the input and output chunks for processing) the 1 km CONUS case takes 1 m 13 s to process and uses 10 GB of memory. The previous iteration used only 2.5 GB, so next I need to look at reducing the memory usage.

**Comment by pnuu** [comment]
The memory consumption is now solved. I was doing a lot of unnecessary padding, stacking and `da.nanmax()`'ing. With the latest commit the ABI case takes 40.0 s to process with a memory use of 2.5 GB. Compared to the 5+ minutes and ~5 GB of the original version in `master` branch, I'm quite happy.

**Comment by pnuu** [comment]
Appveyor doesn't seem to be using conda to install the packages, so I don't know how to fix the linux build error.

**Comment by pnuu** [comment]
I copied some fixes from https://github.com/pytroll/pyresample/pull/247/ so that the arrays passed to Cython code are read-only (`const`) and makes distributed processing to work more reliably. This also required the C source code to be re-generated, which will probably cause a huge merge conflict.

File: pyresample/gradient/__init__.py [hunk 2]
```diff
@@ -51,153 +55,350 @@ def __init__(self, source_geo_def, target_geo_def):
...
+                dst_poly = get_polygon(self.prj, geo_def)
+            except AttributeError:
+                # Can't create polygons for SwathDefinition
+                dst_poly = False
+            self.dst_polys[idx] = dst_poly
+
+        return dst_poly
+
+    def get_chunk_mappings(self):
+        """"""
```
[the remark this answers is not in the document]
Comment (pnuu) [reply pyresample/gradient/__init__.py:137]: Ooops.

File: setup.cfg [hunk 3]
```diff
@@ -1,5 +1,5 @@
 [bdist_rpm]
-requires=python3-numpy pykdtree python3-numexpr pyproj python3-configobj
+requires=python3-numpy pykdtree python3-numexpr pyproj python3-configobj shapely
```
[the remark this answers is not in the document]
Comment (pnuu) [reply setup.cfg]: Updated.

File: pyresample/gradient/__init__.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pnuu) [reply pyresample/gradient/__init__.py]: Docstring added.

**Pull request merged**